src/model/document.ts

```diff
@@ -233,45 +217,66 @@ export class Document implements Feature {
   private _getByKind(kind: string, documentsWalked: Set<Document>):
       Set<Feature> {
     const result = new Set<Feature>();
-    if (documentsWalked.has(this)) {
-      return result;
-    }
     documentsWalked.add(this);
+
     for (const feature of this._localFeatures) {
       if (feature.kinds.has(kind)) {
         result.add(feature);
       }
-      if (feature instanceof Document) {
-        for (const subFeature of feature._getByKind(kind, documentsWalked)) {
-          result.add(subFeature);
+      if (feature.kinds.has('import')) {
+        const document = (feature as Import).document;
+        if (!documentsWalked.has(document)) {
+          for (const subFeature of document._getByKind(kind, documentsWalked)) {
+            result.add(subFeature);
+          }
+        }
+      }
+      if (feature.kinds.has('document')) {
+        const document = (feature as Document);
+        if (!documentsWalked.has(document)) {
+          for (const subFeature of document._getByKind(kind, documentsWalked)) {
+            result.add(subFeature);
+          }
         }
       }
     }
     return result;
   }
 
-  getFeatures(): Set<Feature> {
+  getFeatures(deep: boolean = true): Set<Feature> {
     const result = new Set<Feature>();
-    this._getFeatures(result, new Set<Document>());
+    this._getFeatures(result, new Set<Document>(), deep);
     return result;
   }
 
   private _getFeatures(
-      inProgress: Set<Feature>, documentsWalked: Set<Document>) {
-    if (documentsWalked.has(this)) {
+      result: Set<Feature>, visited: Set<Document>, deep: boolean) {
+    if (visited.has(this)) {
       return;
     }
-    documentsWalked.add(this);
+    visited.add(this);
     for (const feature of this._localFeatures) {
-      inProgress.add(feature);
-      if (feature instanceof Document) {
-        feature._getFeatures(inProgress, documentsWalked);
+      result.add(feature);
+      if (deep) {
+        if (feature.kinds.has('document')) {
```

Comment:
Yeah, but we're using this to access a private API here. I can see the argument for `.has('import')` but it's a bit more fuzzy with something you expect to be type compatible with `this` because it's ok for you to access private members of your own class.

It still might be fine for the really creative plugin writer, but they're also likely to break on minor releases.